packages/dashboard/lib/hooks/useEsbuild.ts

```diff
@@ -0,0 +1,122 @@
+import * as esbuild from 'esbuild-wasm';
+import { Plugin, Loader } from 'esbuild-wasm';
+import { useCallback, useEffect, useState } from 'react';
+
+type EsbuildFileSystem = Map<
+  string,
+  {
+    content: string;
+  }
+>;
+
+const PROJECT_ROOT = '/project/';
+
+const RESOLVE_EXTENSIONS = ['.tsx', '.ts', '.jsx', '.js', '.css', '.json'];
+
+const extname = (path: string): string => {
+  const m = /(\.[a-zA-Z0-9]+)$/.exec(path);
+  return m ? m[1] : '';
+};
+
+const inferLoader = (p: string): Loader => {
+  const ext = extname(p);
+  if (RESOLVE_EXTENSIONS.includes(ext)) {
+    return ext.slice(1) as Loader;
+  }
+  if (ext === '.mjs' || ext === '.cjs') {
+    return 'js';
+  }
+  return 'text';
+};
+
+const resolvePlugin = (files: EsbuildFileSystem): Plugin => {
+  return {
+    name: 'resolve',
+    setup(build) {
+      build.onResolve({ filter: /.*/ }, async args => {
+        if (args.path.startsWith(PROJECT_ROOT)) {
+          return {
+            path: args.path,
+          };
+        }
+      });
+
+      build.onLoad({ filter: /.*/ }, args => {
+        if (args.path.startsWith(PROJECT_ROOT)) {
+          const name = args.path.replace(PROJECT_ROOT, '');
+          const file = files.get(name);
+          if (file) {
+            return {
+              contents: file.content,
+              loader: inferLoader(args.path),
+            };
+          }
+        }
+      });
+    },
+  };
+};
+
+export enum ESBuildStatus {
+  Success,
+  Fail,
+  Loading,
+}
+
+class EsBuildSingleton {
+  static isFirst = true;
+  static getIsFirst = () => {
+    if (EsBuildSingleton.isFirst) {
+      EsBuildSingleton.isFirst = false;
+      return true;
+    }
+    return EsBuildSingleton.isFirst;
+  };
+}
+
+export const useEsbuild = () => {
+  const [esbuildStatus, setEsbuildStatus] = useState(ESBuildStatus.Loading);
+  const [isEsbuildLoading, setIsEsbuildLoading] = useState(true);
+
+  // React.StrictMode will cause useEffect to run twice
+  const loadEsbuild = useCallback(async () => {
+    try {
+      if (EsBuildSingleton.getIsFirst()) {
+        await esbuild.initialize({
+          wasmURL: `/esbuild.wasm`,
```

Comment:
I put it in the `public` folder from our `node_modules`, as I don't know the CDN addresses for each area. So I chose to do it this way.

Reply:
Let's remove it from the `public` folder and instead download the wasm binary from `https://esm.sh/esbuild-wasm@0.18.7/esbuild.wasm`. This will avoid potentially high load and bandwidth usage on our side, which is not something we want.

Reply:
done